core/src/main/java/com/linecorp/armeria/common/stream/FixedStreamMessage.java

```diff
@@ -133,32 +132,41 @@ final void notifySubscriberOfCloseEvent(SubscriptionImpl subscription, CloseEven
 
     @Override
     final void cancel() {
-        cancelOrAbort(true);
+        cancelOrAbort(CancelledSubscriptionException.get());
     }
 
     @Override
     public final void abort() {
+        abort0(AbortedStreamException.get());
+    }
+
+    @Override
+    public final void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        abort0(cause);
+    }
+
+    private void abort0(Throwable cause) {
         final SubscriptionImpl currentSubscription = subscription;
         if (currentSubscription != null) {
-            cancelOrAbort(false);
+            cancelOrAbort(cause);
             return;
         }
 
         final SubscriptionImpl newSubscription = new SubscriptionImpl(
-                this, AbortingSubscriber.get(), ImmediateEventExecutor.INSTANCE, false, false);
+                this, AbortingSubscriber.get(cause), ImmediateEventExecutor.INSTANCE, false, false);
         subscriptionUpdater.compareAndSet(this, null, newSubscription);
-        cancelOrAbort(false);
+        cancelOrAbort(cause);
     }
 
-    private void cancelOrAbort(boolean cancel) {
+    private void cancelOrAbort(Throwable cause) {
         final CloseEvent closeEvent;
-        final Sampler<Class<? extends Throwable>> sampler = Flags.verboseExceptionSampler();
-        if (cancel) {
-            closeEvent = sampler.isSampled(CancelledSubscriptionException.class) ?
-                         new CloseEvent(new CancelledSubscriptionException()) : CANCELLED_CLOSE;
+        if (cause == CancelledSubscriptionException.INSTANCE) {
+            closeEvent = CANCELLED_CLOSE;
+        } else if (cause == AbortedStreamException.INSTANCE) {
+            closeEvent = ABORTED_CLOSE;
         } else {
-            closeEvent = sampler.isSampled(AbortedStreamException.class) ?
-                         new CloseEvent(new AbortedStreamException()) : ABORTED_CLOSE;
+            closeEvent = new CloseEvent(cause);
         }
```

Comment:
I see similar logic in a few places. Can we move this common logic to a static method in `AbstreamStreamMessage`?

Reply:
Sure, I will address it!